packages/functional/src/__tests__/pipe-test.ts

```diff
@@ -0,0 +1,382 @@
+import { pipe } from '../pipe';
+
+describe('pipe', () => {
+    it('can pipe a single value', () => {
+        expect.assertions(6);
+        expect(pipe(true)).toBe(true);
+        expect(pipe('test')).toBe('test');
+        expect(pipe(1)).toBe(1);
+        expect(pipe(3n)).toBe(3n);
+        expect(pipe(null)).toBeNull();
+        expect(pipe(undefined)).toBeUndefined();
+    });
+    it('can pipe a single function', () => {
+        expect.assertions(1);
+        expect(pipe('test', value => value.toUpperCase())).toBe('TEST');
+    });
+    it('can pipe multiple functions', () => {
+        expect.assertions(3);
+        expect(
+            pipe(
+                'test',
+                value => value.toUpperCase(),
+                value => value + '!',
+                value => value.repeat(3)
+            )
+        ).toBe('TEST!TEST!TEST!');
+        expect(
+            pipe(
+                1,
+                value => value + 1,
+                value => value + 2,
+                value => value + 3
+            )
+        ).toBe(7);
+        expect(
+            pipe(
+                1,
+                value => value + 1,
+                value => value * 2,
+                value => value - 1
+            )
+        ).toBe(3);
+    });
+    it('can pipe multiple functions with different types', () => {
+        expect.assertions(3);
+        expect(
+            pipe(
+                1,
+                value => value + 1,
+                value => value.toString(),
+                value => value + '!'
+            )
+        ).toBe('2!');
+        expect(
+            pipe(
+                'test',
+                value => value.toUpperCase(),
+                value => value.length,
+                value => value + 1
+            )
+        ).toBe(5);
+        expect(
+            pipe(
+                1,
+                value => value + 1,
+                value => value * 2,
+                value => value - 1,
+                value => value.toString(),
+                value => value + '!'
+            )
+        ).toBe('3!');
+    });
+    describe('combining objects', () => {
+        function combine<T extends object, U extends object>(a: T, b: U): T & U {
+            return { ...a, ...b };
+        }
+        it('can combine two objects', () => {
+            expect.assertions(1);
+            expect(pipe({ a: 1 }, value => combine(value, { b: 2 }))).toEqual({ a: 1, b: 2 });
+        });
```

Comment:
The only _possible_ addition that I could make here would be a test that asserts that there's no copy-funny-business happening inside `pipe` itself. Like I'd want to assert that the object I put in is referentially equal to the one that comes out.

```ts
const startObj = {};
const endObj = pipe(
  startObj,
  (obj) => {
    obj.hello = 'there'; // A mutation of the passed-in-object
    return obj;
  },
);
expect(startObj).toBe(endObj);
```

Reply:
I actually can't even set this test up without TypeScript throwing

```
Property 'hello' does not exist on type '{}'
```

But I did add 2 tests to address your concerns!